canned/postgresql.go

```diff
@@ -0,0 +1,147 @@
+package canned
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net"
+	"os"
+	"syscall"
+	"time"
+
+	_ "github.com/lib/pq"
+	"github.com/pkg/errors"
+	testcontainers "github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+const (
+	user       = "user"
+	password   = "password"
+	database   = "database"
+	image      = "postgres"
+	defaultTag = "11.5"
+	port       = "5432/tcp"
+)
+
+type PostgreSQLContainerRequest struct {
+	testcontainers.GenericContainerRequest
+	User     string
+	Password string
+	Database string
+}
+
+type postgresqlContainer struct {
+	Container testcontainers.Container
+	db        *sql.DB
+	req       PostgreSQLContainerRequest
+}
+
+func (c *postgresqlContainer) GetDriver(ctx context.Context) (*sql.DB, error) {
+
+	host, err := c.Container.Host(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mappedPort, err := c.Container.MappedPort(ctx, port)
+	if err != nil {
+		return nil, err
+	}
+
+	db, err := sql.Open("postgres", fmt.Sprintf(
+		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
+		host,
+		mappedPort.Int(),
+		c.req.User,
+		c.req.Password,
+		c.req.Database,
+	))
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+func PostgreSQLContainer(ctx context.Context, req PostgreSQLContainerRequest) (*postgresqlContainer, error) {
+
+	provider, err := req.ProviderType.GetProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// With the current logic it's not really possible to allow other ports...
+	req.ExposedPorts = []string{port}
+
+	if req.Env == nil {
+		req.Env = map[string]string{}
+	}
+
+	// Set the default values if none were provided in the request
+	if req.Image == "" {
+		req.Image = fmt.Sprintf("%s:%s", image, defaultTag)
+	}
+
+	if req.User == "" {
+		req.User = user
+	}
+
+	if req.Password == "" {
+		req.Password = password
+	}
+
+	if req.Database == "" {
+		req.Database = database
+	}
+
+	req.Env["POSTGRES_USER"] = req.User
+	req.Env["POSTGRES_PASSWORD"] = req.Password
+	req.Env["POSTGRES_DB"] = req.Database
+
+	req.WaitingFor = wait.ForAll(
+		wait.ForLog("database system is ready to accept connections"),
+		wait.ForListeningPort(port),
+	)
+
+	c, err := provider.CreateContainer(ctx, req.ContainerRequest)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create container")
+	}
+
+	postgresC := &postgresqlContainer{
+		Container: c,
+		req:       req,
+	}
+
+	if req.Started {
+		if err := c.Start(ctx); err != nil {
+			return postgresC, errors.Wrap(err, "failed to start container")
+		}
+
+		db, err := postgresC.GetDriver(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		for {
```

Comment:
I guess this is because the postgres container "bounces" and logs the "database system is ready to accept connections" twice?

In the java testcontainer that's handled by having the Wait code have a state and wait for it twice.

I think I prefer this, as it keeps the Waits simple

Reply:
Yes, I don't understand why the Postgres container is doing that, but basically the process gets restarted, so you'd have to wait for the second `"database system is ready to accept connections"`.

Because of that though I realised, the wait strategy I put there is basically pointless? :thinking: 

I have another proposal to make the same wait process into a proper `SQLStrategy` wait strategy.